progress/progress.go

```diff
@@ -0,0 +1,192 @@
+package progress
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/lucasb-eyer/go-colorful"
+	"github.com/muesli/reflow/ansi"
+	"github.com/muesli/termenv"
+)
+
+const defaultWidth = 40
+
+var color func(string) termenv.Color = termenv.ColorProfile().Color
+
+// Option is used to set options in NewModel. For example:
+//
+//     progress := NewModel(
+//	       WithRamp("#ff0000", "#0000ff"),
+//	       WithoutPercentage(),
+//     )
+type Option func(*Model) error
+
+// WithDefaultRamp sets a gradient fill with default colors.
+func WithDefaultRamp() Option {
+	return WithRamp("#00dbde", "#fc00ff")
+}
+
+// WithRamp sets a gradient fill blending between two colors.
+func WithRamp(colorA, colorB string) Option {
+	return func(m *Model) error {
+		return m.setRamp(colorA, colorB, false)
+	}
+}
+
+// WithDefaultScaledRamp sets a gradient with default colors, and scales the
+// gradient to fit the filled portion of the ramp.
+func WithDefaultScaledRamp() Option {
+	return WithScaledRamp("#00dbde", "#fc00ff")
+}
+
+// WithScaledRamp scales the gradient to fit the width of the filled portion of
+// the progress bar.
+func WithScaledRamp(colorA, colorB string) Option {
+	return func(m *Model) error {
+		return m.setRamp(colorA, colorB, true)
+	}
+}
+
+// WithSolidFill sets the progress to use a solid fill with the given color.
+func WithSolidFill(color string) Option {
+	return func(m *Model) error {
+		m.FullColor = color
+		m.useRamp = false
+		return nil
+	}
+}
+
+// WithoutPercentage hides the numeric percentage.
+func WithoutPercentage() Option {
+	return func(m *Model) error {
+		m.ShowPercentage = false
+		return nil
+	}
+}
+
+// WithWidth sets the initial width of the progress bar. Note that you can also
+// set the width via the Width property, which can come in handy if you're
+// waiting for a tea.WindowSizeMsg.
+func WithWidth(w int) Option {
+	return func(m *Model) error {
+		m.Width = w
+		return nil
+	}
+}
+
+// Model stores values we'll use when rendering the progress bar.
+type Model struct {
+
+	// Total width of the progress bar, including percentage, if set.
+	Width int
+
+	// "Filled" sections of the progress bar
+	Full      rune
+	FullColor string
+
+	// "Empty" sections of progress bar
+	Empty      rune
+	EmptyColor string
+
+	// Settings for rendering the numeric percentage
+	ShowPercentage  bool
+	PercentFormat   string // a fmt string for a float
+	PercentageStyle *termenv.Style
+
+	useRamp    bool
```

Comment:
maybe we should make it public?

Reply:
I'm hesitant to make it public because it also requires generating the two `colorful` colors that go with it. In that case we could use a setter function, but I figure at that point one may as well just recreate the struct with `NewModel()`.